### ISSUE TITLE
eventing_epp custom metrics has only collect successful transmit with response_code as 204 when destination is nats backend

### ISSUE THREAD

**Issue by vvxxvvxx** [kyma-project/kyma#15936]
**Description**

For the below 3 eventing_epp custom metrics with label `response_code`, for the case that destination_service is nats backend, in the codes, only response_code 204 will be recorded. All the erroneous response cases are disregarded. 
We need to collect all kinds of responses and classify them to 2xx for successful case and 5xx for failed case.
```
eventing_epp_messaging_server_latency_duration_milliseconds
epp_event_type_published_total
eventing_epp_requests_total
```

As we can see, for nats backend, the Send(..) returns a fixed response_code which is 204 for the successful sending out case.
But what if the event is not sent successfully? for the erroneous case, it is not considered in the codes.
https://github.com/kyma-project/kyma/blob/608d724bcb66e39251cc6ae2efb02245eaf47b0c/components/event-publisher-proxy/pkg/sender/jetstream/jetstream.go#L91

But we can find the codes from nats.go, for example, the codes here https://github.com/nats-io/nats.go/blob/ffbe2f99dcc9987100cbf1fad13535a68dcf7938/nats.go#L3771, it gets the case when the response header indicates `noResponders`, then `ErrNoResponders` will be returned. but in our codes, we ignore all such errors, and only return the good case using response_code as 204.

We also need to collect the erroneous cases or errors from response, and then define the errors for response_code as 5xx. Then the metrics will be more meaningful. We can then make use of different response_code values to consider the availability.

And in the codes https://github.com/kyma-project/kyma/blob/608d724bcb66e39251cc6ae2efb02245eaf47b0c/components/event-publisher-proxy/pkg/handler/handler.go#L191, don't return `return nil, err` directly when `err!=nil` from `result, err := h.Sender.Send(ctx, event)`, instead, collect the metrics for the erroneous cases. because as long as we receive the response from the nats-server, the transmit happened, we need to record the latency and response_code for them instead of only recording the successful request-response.
https://github.com/kyma-project/kyma/blob/608d724bcb66e39251cc6ae2efb02245eaf47b0c/components/event-publisher-proxy/pkg/handler/handler.go#L196

**Expected result**
The metrics collect all kinds of responses and classify them to 2xx for successful case and 5xx for failed case.

**Actual result**
Only successful transmit is collected in metrics for events sent to nats backend.

**Steps to reproduce**

**Troubleshooting**

**Comment by k15r** [comment]
Tasks:
- always record metrics if interacting with the backend

**Comment by vvxxvvxx** [comment]
Hi, with the current implementation, seems this issue (`eventing_epp_backend_duration_milliseconds` always responds 204 response code) still exists.

in this line of codes https://github.com/kyma-project/kyma/blob/066180f11902674c331866e6c2e23b58ca554641/components/event-publisher-proxy/pkg/handler/handler.go#L320, we can see if err!=nil, it will return. 
and in the implementation of Sender(..), if it's jetstream, then only http.StatusNoContent is returned when err==nil https://github.com/kyma-project/kyma/blob/066180f11902674c331866e6c2e23b58ca554641/components/event-publisher-proxy/pkg/sender/jetstream/jetstream.go#L102. when err!=nil, this returned sender.PublishResult, is nil.
Can we also define the corresponding 5xx or 4xx response code when err!=nil in Sender(..)?
and in handler.go, don't directly return when err!=nil. then 5xx/4xx/2xx these response code all can be captured in below metrics. 
https://github.com/kyma-project/kyma/blob/066180f11902674c331866e6c2e23b58ca554641/components/event-publisher-proxy/pkg/handler/handler.go#L335
```
	h.collector.RecordEventType(originalEventType, event.Source(), result.HTTPStatus())
	h.collector.RecordBackendLatency(duration, result.HTTPStatus(), host)
	h.collector.RecordBackendRequests(result.HTTPStatus(), host)
```

**Comment by friedrichwilken** [comment]
@k15r can you give update what the latest state here is?

**Comment by k15r** [comment]
There is a mapping of Jetstream internal errors to http error codes somewhere in the jetstream codebase. This change will require a partial rewrite of our own error handling in the epp.

**Comment by grischperl** [comment]
Keeping this closed as #17547 will address this issue (and already has 2 PRs linked), has more current information as well as pointing back here as the point of origin.
